lib/oxide-vpc/src/api.rs

```diff
@@ -20,13 +20,101 @@ use serde::Deserialize;
+    #[default]
+    External = 0x00,
+    /// Replicate packets to underlay/infrastructure members.
+    ///
+    /// Forwards Geneve-encapsulated packets to underlay destinations for
+    /// infrastructure delivery (not directly to guest instances). Uses
+    /// DEFAULT_MULTICAST_VNI (77) for encapsulation.
+    Underlay = 0x01,
+    /// Replicate packets to both external and underlay members (bifurcated).
+    ///
+    /// Combines both customer-facing (decapsulated to guests) and infrastructure
+    /// (encapsulated) delivery modes for comprehensive multicast distribution.
+    All = 0x02,
+    /// Reserved for future use. This value exists to account for all possible
+    /// values in the 2-bit Geneve option field.
+    Reserved = 0x03,
+}
+
+impl Replication {
+    /// Merge two replication strategies, preferring the most permissive.
+    ///
+    /// Merging rules:
+    /// - Any `All` -> `All`
+    /// - `External` + `Underlay` -> `All`
+    /// - Same values -> keep the value
+    /// - Default to `All` for unexpected combinations
+    pub const fn merge(self, other: Self) -> Self {
+        match (self, other) {
+            (Self::All, _) | (_, Self::All) => Self::All,
+            (Self::External, Self::Underlay)
+            | (Self::Underlay, Self::External) => Self::All,
+            (a, b) if a as u8 == b as u8 => a,
+            // Prefer `All` for unexpected combinations
+            _ => Self::All,
+        }
+    }
+}
+
+#[cfg(any(feature = "std", test))]
+impl FromStr for Replication {
+    type Err = String;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match s.to_ascii_lowercase().as_str() {
+            "external" => Ok(Self::External),
+            "underlay" => Ok(Self::Underlay),
+            "all" => Ok(Self::All),
+            lower => Err(format!(
+                "unexpected replication type {lower} -- expected 'external', 'underlay', or 'all'"
+            )),
+        }
+    }
+}
+
 /// This is the MAC address that OPTE uses to act as the virtual gateway.
 pub const GW_MAC_ADDR: MacAddr =
     MacAddr::from_const([0xA8, 0x40, 0x25, 0xFF, 0x77, 0x77]);
 /// The default VNI ID which OPTE uses for outbound packets directed at a
 /// tunnel endpoint.
 pub const BOUNDARY_SERVICES_VNI: u32 = 99u32;
 
+/// Default VNI for rack-wide multicast groups (no VPC association).
+/// Must match Omicron's DEFAULT_MULTICAST_VNI.
+///
+/// This is the only VNI currently supported for multicast traffic.
+/// All multicast groups (M2P mappings and forwarding entries) must use this VNI.
+/// OPTE validates that multicast operations specify this VNI and rejects others.
+///
+/// **Security model:** While M2P (Multicast-to-Physical) mappings are stored
+/// per-VNI in the code, the enforcement of DEFAULT_MULTICAST_VNI means all
+/// multicast traffic shares a single namespace across the rack, with no
+/// VPC-level isolation (as multicast groups are fleet-wide).
+pub const DEFAULT_MULTICAST_VNI: u32 = 77u32;
```

Comment:
Okay, so we aren't yet doing any VPC-level isolation. I guess the only thing we need to be aware of is what the source IP looks like, IIRC there was discussion at some point that in this case all packets should be NAT'd.